addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java

```diff
@@ -0,0 +1,321 @@
+                return;
+            }
+
+            State state = null;
+            if (value == null) {
+                state = UnDefType.UNDEF;
+            } else if (value instanceof Calendar) {
+                state = new DateTimeType((Calendar) value);
+            } else if (value instanceof BigDecimal) {
+                state = new DecimalType((BigDecimal) value);
+            } else if (value instanceof Integer) {
+                state = new DecimalType(BigDecimal.valueOf(((Integer) value).longValue()));
+            } else if (value instanceof String) {
+                state = new StringType(value.toString());
+            } else {
+                logger.warn("Update channel {}: Unsupported value type {}", channelId,
+                        value.getClass().getSimpleName());
+            }
+            logger.debug("Update channel {} with state {} ({})", channelId, (state == null) ? "null" : state.toString(),
+                    (value == null) ? "null" : value.getClass().getSimpleName());
+
+            // Update the channel
+            if (state != null) {
+                updateState(channelId, state);
+            }
+        }
+    }
+
+    /**
+     * Get new data from aqicn.org service
+     *
+     * @return {boolean}
+     */
+    private AirQualityJsonResponse updateAirQualityData() {
+        AirQualityConfiguration config = getConfigAs(AirQualityConfiguration.class);
+        return getAirQualityData(StringUtils.trimToEmpty(config.getLocation()), config.getStationId());
+    }
+
+    /**
+     * Request new air quality data to the aqicn.org service
+     *
+     * @param location geo-coordinates from config
+     * @param stationId station ID from config
+     *
+     * @return the air quality data object mapping the JSON response or null in case of error
+     */
+    private AirQualityJsonResponse getAirQualityData(String location, Integer stationId) {
+        AirQualityJsonResponse result = null;
+        boolean resultOk = false;
+        String errorMsg = null;
+
+        try {
+
+            // Build a valid URL for the aqicn.org service
+            AirQualityConfiguration config = getConfigAs(AirQualityConfiguration.class);
+
+            String geoStr = "geo:";
+            geoStr += location.replace(" ", "").replace(",", ";").replace("\"", "").replace("'", "").trim();
+
+            String urlStr = URL.replace("%APIKEY%", StringUtils.trimToEmpty(config.getApiKey()));
+
+            if (stationId == null) {
+                urlStr = urlStr.replace("%QUERY%", geoStr);
+            } else {
+                urlStr = urlStr.replace("%QUERY%", "@" + stationId);
+            }
+
+            logger.debug("URL = {}", urlStr);
+
+            // Run the HTTP request and get the JSON response from aqicn.org
+            URL url = new URL(urlStr);
+            URLConnection connection = url.openConnection();
+            String repsonse = IOUtils.toString(connection.getInputStream());
```

Comment:
Better do a try-with-resources as [IOUtls.toString does not close the input stream by itself](https://commons.apache.org/proper/commons-io/description.html).